### PR TITLE
Make detach redispatch like a regular PyTorch operator

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71707

Why?
- detach should behave like jax.stop_gradient in functorch. Because it
does not detach all the way through, functorch (as well as a Tensor
Subclass wrapping a Tensor subclass) won't see it after the first
layer/subclass handles it.

How?
- This PR changes detach to dispatch all the way through to the backend.
- This PR also modifies native::detach to call shallow_copy_and_detach
instead of native::alias. This is because today, the semantics of detach
and alias are differently -- they differ only by
allow_tensor_metadata_change. In the future, we may choose to deprecate
this flag.
- NB: Before and after this PR, detach() shows up twice in
torch_dispatch: https://github.com/pytorch/pytorch/issues/71725. This is
not a regression so I didn't want to fix it in this PR because it is
weird to fix.

Test Plan:
- added new tests; run existing tests

Differential Revision: [D33752860](https://our.internmc.facebook.com/intern/diff/D33752860)

### CODE REVIEW

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33752860).

File: test/test_autograd.py [hunk 1]
```diff
@@ -2306,6 +2306,16 @@ def test_detach_base(self):
         self.assertIsNotNone(view.grad_fn)
         self.assertIs(view._base, x)
 
+    def test_detach_then_inplace_raises_in_autograd(self):
```
Comment (albanD) [inline test/test_autograd.py:2309]: What is this testing?
If you try to test the metadata change, you want to do `z.resize_()` or `z.as_strided_()` (any inplace view) and these should fail when called.

File: test/test_python_dispatch.py [hunk 2]
```diff
@@ -137,9 +137,23 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
         )
         self.assertExpectedRaisesInline(
             RuntimeError, lambda: A(torch.zeros(1)).detach(),
-            """detach returned invalid type str, expected Tensor"""
+            """Unable to cast Python instance of type <class 'str'> to C++ type 'at::Tensor'"""
```
Comment (albanD) [inline test/test_python_dispatch.py:140]: This error message is not as nice as it used to be... Why is that?

File: test/test_autograd.py (hunk 1 above)
Comment (zou3519) [reply test/test_autograd.py:2309]: I'll add an additional test for the metadata change, thanks for that suggestion

I think this test is testing that the version counter gets propagated? In a previous version of this PR, this test did not succeed

File: test/test_python_dispatch.py (hunk 2 above)
Comment (zou3519) [reply test/test_python_dispatch.py:140]: Previously, this error message came from shallow_copy_and_detach, which calls concrete_detach_fn, which hardcodes the error message:
https://github.com/pytorch/pytorch/blob/f5a71ec2d6956a1ba393657f4b8297bb931eeec2/torch/csrc/autograd/python_variable.cpp#L1828

After this PR, the first detach that gets called is not one from shallow_copy_and_detach. This one does not go through `concrete_detach_fn` so it goes through the usual pybind error reporting mechanism and pybind is very undescriptive (see the error message that is two lines up from this one)

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py:2309]: The version counter should already be properly propagated in master. Maybe there is already a test for that? But if not, we definitely want one.

Comment (zou3519) [reply test/test_autograd.py:2309]: There isn't a test for it (for detach). The version counter is indeed properly propagated in master

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33752860).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33752860).
